sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCpuBridgeOptimizer.scala

```diff
@@ -0,0 +1,315 @@
+ * and combining cpu expressions into an expression tree.
+ */
+object GpuCpuBridgeOptimizer extends Logging {
+
+  def checkAndOptimizeExpressionMetas(exprs: Seq[BaseExprMeta[_]]): Unit = {
+    if (exprs.nonEmpty && exprs.head.conf.isCpuBridgeEnabled) {
+      exprs.foreach { child =>
+        if (!child.canExprTreeBeReplaced && canRunOnCpuOrGpuRecursively(child)) {
+          // Then we want to move some things to the CPU as needed
+          moveToCpuIfNeededRecursively(child, wasParentReplaced = false, 0)
+        }
+      }
+    }
+  }
+
+  private def canRunOnCpuOrGpuRecursively(expr: BaseExprMeta[_]): Boolean = {
+    if (!expr.canThisBeReplaced && !expr.canMoveToCpuBridge()) {
+      false
+    } else {
+      expr.childExprs.forall(canRunOnCpuOrGpuRecursively)
+    }
+  }
+
+  /**
+   * Move an expression to the GPU/CPU bridge if it has to or if it would be
+   * less data movement if it did run on the CPU, this means that it has more inputs
+   * and outputs on the CPU than the GPU.
+   * @param expr the expression to possibly move
+   * @param wasParentReplaced will its output be on the CPU
+   * @return true if this was placed on the CPU bridge else false
+   */
+  private def moveToCpuIfNeededRecursively(expr: BaseExprMeta[_],
+                                           wasParentReplaced: Boolean,
+                                           indent: Int): Boolean = {
+    if (expr.willUseGpuCpuBridge) {
+      // Some expression trees can be reused. If we are here, then we don't need to go any
+      // deeper
+      true
+    } else if (!expr.canThisBeReplaced && expr.canMoveToCpuBridge()) {
+      expr.childExprs.foreach { child =>
+        moveToCpuIfNeededRecursively(child, wasParentReplaced = true, indent + 1)
+      }
+      expr.moveToCpuBridge()
+      true
+    } else if (expr.canThisBeReplaced && expr.canMoveToCpuBridge()) {
+      // We need to check the cost of moving this (data movement only for now)
+      // But we don't know if we are going to move until we know our children will
+      // But they need to know if we will before they can tell, so just assume we
+      // will not move and the children may be wrong if that is incorrect, but
+      // we want to prefer the GPU for execution if possible
+      val maxPossibleCost: Int = expr.childExprs.length + 1
+      // This is for the output
+      val parentCost = if (wasParentReplaced) 1 else 0
+      val childrenMovedToCpu = expr.childExprs.map { child =>
+        moveToCpuIfNeededRecursively(child, wasParentReplaced = false, indent + 1)
+      }
+      val costToStayOnGPU = childrenMovedToCpu.count(b => b) + parentCost
+      val costToMoveDataToCPU = maxPossibleCost - costToStayOnGPU
+
+      // prefer the GPU if things are equal
+      if (costToStayOnGPU > costToMoveDataToCPU) {
+        expr.moveToCpuBridge()
+        true
+      } else {
+        false
+      }
+    } else {
+      // This can be replaced, but cannot run on the bridge, so check its children
```

Comment:
Is it actually guaranteed that `expr.canThisBeReplaced` is true? To me it looks like `if !expr.canThisBeReplaced && !expr.canMoveToCpuBridge()` we should still end up at this branch, but maybe there is a further assumption I'm missing.